syntax = "proto3";

package vad;

// VAD Service Definition
service VADService {
  // Create a new VAD session
  rpc CreateSession(CreateSessionRequest) returns (CreateSessionResponse);
  
  // Detect voice activity in audio data
  rpc Detect(DetectRequest) returns (DetectResponse);
  
  // Reset session state (clear buffers)
  rpc ResetSession(ResetSessionRequest) returns (Empty);
  
  // Destroy a VAD session
  rpc DestroySession(DestroySessionRequest) returns (Empty);
  
  // Get service stats
  rpc GetStats(Empty) returns (StatsResponse);
}

// Request to create a new session
message CreateSessionRequest {
  string session_id = 1;
  float threshold = 2;
  float threshold_low = 3;
  int32 frame_window_threshold = 4;
}

// Response for session creation
message CreateSessionResponse {
  bool success = 1;
  string message = 2;
}

// Request to detect voice activity
message DetectRequest {
  string session_id = 1;
  bytes audio_data = 2;  // PCM 16kHz 16-bit mono
  string event = 3;      // "start" / "chunk" / "end"
}

// Response with detection result
message DetectResponse {
  bool has_voice = 1;
  bool is_speaking = 2;
  string session_id = 3;
}

// Request to reset session
message ResetSessionRequest {
  string session_id = 1;
}

// Request to destroy session
message DestroySessionRequest {
  string session_id = 1;
}

// Empty message
message Empty {
}

// Service statistics
message StatsResponse {
  int32 active_sessions = 1;
  int64 total_requests = 2;
  repeated SessionStats sessions = 3;
}

message SessionStats {
  string session_id = 1;
  int64 request_count = 2;
  double age_seconds = 3;
  bool vad_locked = 4;
}

